From 81152c8345bc291df4c76c3463e1ff8e84876da1 Mon Sep 17 00:00:00 2001
From: kikairoya <kikairoya@gmail.com>
Date: Tue, 24 Mar 2026 07:30:26 +0900
Subject: [PATCH 08/35] [Clang][MinGW] Instantiate nested classes by enclosing
 template's explicit instantiation declaration

In-code comment says "explicit instantiation decl of the outer class
doesn't affect the inner class" but this behavior seems MSVC
specific, MinGW-GCC and Cygwin-GCC does not.
Clang should honor gcc's behavior.

This change fixes std::string compatibilty and resolves strange link
error (statically linked), strange crash (dynamically linked) using
libstdc++ on Cygwin.

[Clang][MinGW] Export nested classes which instantiated by enclosing template's explicit instantiation declaration

MinGW-GCC applies explicit instantiateion declaration of a class template to its nested type but doesn't inherit
`__declspec(dllexport)` attribute to the nested class, causes undefined reference to member functions of the nested
class.
To resolve this self-inconsistency, such `__declspec(dllexport)` attribute on explicit instantiation declaration should
inherit to the nested class.
---
 clang/lib/Sema/SemaDeclCXX.cpp             | 21 +++++++++++++++++++--
 clang/lib/Sema/SemaTemplateInstantiate.cpp |  1 +
 2 files changed, 20 insertions(+), 2 deletions(-)

diff --git a/clang/lib/Sema/SemaDeclCXX.cpp b/clang/lib/Sema/SemaDeclCXX.cpp
index 133e7df05ba4c9f4d0a662d3719b7f0bcecd69e3..06457223f952af24fe86bfbceba7968a76b484a4 100644
--- a/clang/lib/Sema/SemaDeclCXX.cpp
+++ b/clang/lib/Sema/SemaDeclCXX.cpp
@@ -6672,12 +6672,13 @@ void Sema::checkClassLevelDLLAttribute(CXXRecordDecl *Class) {
       cast<DLLImportAttr>(ClassAttr)->wasPropagatedToBaseTemplate();
 
   TemplateSpecializationKind TSK = Class->getTemplateSpecializationKind();
+  const TargetInfo &TI = Context.getTargetInfo();
 
   // Ignore explicit dllexport on explicit class template instantiation
   // declarations, except in MinGW mode.
   if (ClassExported && !ClassAttr->isInherited() &&
       TSK == TSK_ExplicitInstantiationDeclaration &&
-      !Context.getTargetInfo().getTriple().isOSCygMing()) {
+      !TI.getTriple().isOSCygMing()) {
     if (auto *DEA = Class->getAttr<DLLExportAttr>()) {
       Class->addAttr(DLLExportOnDeclAttr::Create(Context, DEA->getLoc()));
       Class->dropAttr<DLLExportAttr>();
@@ -6721,14 +6722,30 @@ void Sema::checkClassLevelDLLAttribute(CXXRecordDecl *Class) {
 
     VarDecl *VD = dyn_cast<VarDecl>(Member);
     CXXMethodDecl *MD = dyn_cast<CXXMethodDecl>(Member);
+    CXXRecordDecl *RD = dyn_cast<CXXRecordDecl>(Member);
     TemplateSpecializationKind MemberTSK =
         VD   ? VD->getTemplateSpecializationKind()
         : MD ? MD->getTemplateSpecializationKind()
+        : RD ? RD->getTemplateSpecializationKind()
              : TSK_Undeclared;
 
     if (MemberTSK == TSK_ExplicitSpecialization)
       continue;
 
+    // Inherit dllexport to nested class in MinGW mode.
+    if (RD) {
+      if (TI.getTriple().isOSCygMing() && !getDLLAttr(Member) &&
+          (TSK == TSK_ExplicitInstantiationDeclaration ||
+           TSK == TSK_ExplicitInstantiationDefinition) &&
+          isTemplateInstantiation(MemberTSK)) {
+        auto *NewAttr =
+            cast<InheritableAttr>(ClassAttr->clone(getASTContext()));
+        NewAttr->setInherited(true);
+        Member->addAttr(NewAttr);
+      }
+      continue;
+    }
+
     // Only methods and static fields inherit the attributes.
     if (!VD && !MD)
       continue;
@@ -6792,7 +6809,7 @@ void Sema::checkClassLevelDLLAttribute(CXXRecordDecl *Class) {
         // MinGW does not import or export inline methods. But do it for
         // template instantiations and inherited constructors (which are
         // marked inline but must be exported to match MSVC behavior).
-        if (!Context.getTargetInfo().shouldDLLImportComdatSymbols() &&
+        if (!TI.shouldDLLImportComdatSymbols() &&
             TSK != TSK_ExplicitInstantiationDeclaration &&
             TSK != TSK_ExplicitInstantiationDefinition) {
           if (auto *CD = dyn_cast<CXXConstructorDecl>(MD);
diff --git a/clang/lib/Sema/SemaTemplateInstantiate.cpp b/clang/lib/Sema/SemaTemplateInstantiate.cpp
index 06ea12ca7bd73bd77d2c219ab7f872dd44dc7bdb..f0b4821022ced34d004ff36d61ce875e646944b8 100644
--- a/clang/lib/Sema/SemaTemplateInstantiate.cpp
+++ b/clang/lib/Sema/SemaTemplateInstantiate.cpp
@@ -4329,6 +4329,7 @@ Sema::InstantiateClassMembers(SourceLocation PointOfInstantiation,
         continue;
 
       if (Context.getTargetInfo().getTriple().isOSWindows() &&
+          !Context.getTargetInfo().getTriple().isOSCygMing() &&
           TSK == TSK_ExplicitInstantiationDeclaration) {
         // On Windows, explicit instantiation decl of the outer class doesn't
         // affect the inner class. Typically extern template declarations are
-- 
2.55.0

